# 📘 Pre-entrega

**Nombre del alumno:**  ***Lorena Zolss***

### 1️⃣ Carga de datos

In [37]:
import pandas as pd
import numpy as np

In [38]:
# Monto mi drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [39]:
# Verifico que los archivos csv se encuentren en la carpeta datasets
import os
os.listdir('/content/drive/MyDrive/datasets')

['ventas.sqlite',
 'users.csv',
 'reviews.csv',
 'Kaggle_users.xlsx',
 'hotels.csv',
 'clientes.csv',
 'marketing.csv',
 'ventas.csv',
 'ventas_clean.csv',
 'clientes_clean.csv',
 'marketing_clean.csv']

In [40]:
# Defino las rutas de los datasets y los cargo como dataframes
ruta_ventas = '/content/drive/MyDrive/datasets/ventas.csv'
ruta_marketing = '/content/drive/MyDrive/datasets/marketing.csv'
ruta_clientes = '/content/drive/MyDrive/datasets/clientes.csv'

df_ventas = pd.read_csv(ruta_ventas)
df_marketing = pd.read_csv(ruta_marketing)
df_clientes = pd.read_csv(ruta_clientes)

In [41]:
# Compruebo que se hayan cargado correctamente.
print("\n****** Archivo ventas.csv ******")
print("Shape :",df_ventas.shape)
display(df_ventas.head())
print()
print("_"*80)
print("\n****** Archivo marketing.csv ******")
print("Shape :",df_marketing.shape)
display(df_marketing.head())
print()
print("_"*80)
print("\n****** Archivo clientes.csv ******")
print("Shape :",df_clientes.shape)
display(df_clientes.head())
print("_"*80)


****** Archivo ventas.csv ******
Shape : (3035, 6)


id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro decorativo   $69.94       5.0  02/01/2024   
1       811    Lámpara de mesa  $105.10       5.0  02/01/2024   
2      1156           Secadora   $97.96       3.0  02/01/2024   
3      1372           Heladera  $114.35       8.0  02/01/2024   
4      1546           Secadora  $106.21       4.0  02/01/2024   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos


________________________________________________________________________________

****** Archivo marketing.csv ******
Shape : (90, 6)


id_campanha         producto  canal  costo fecha_inicio   fecha_fin
0           74  Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12           Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32  Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21       Smartphone   RRSS   6.37   29/03/2024  16/05/2024
4           58         Alfombra  Email   4.25   31/03/2024  05/05/2024


________________________________________________________________________________

****** Archivo clientes.csv ******
Shape : (567, 5)


id_cliente               nombre  edad         ciudad  ingresos
0           1      Aloysia Screase    44  Mar del Plata  42294.68
1           2  Kristina Scaplehorn    25        Posadas  24735.04
2           3       Filip Castagne    50    Resistencia  35744.85
3           4          Liuka Luard    39   Bahía Blanca  27647.96
4           5        Dore Cockshtt    28        Rosario  28245.65

________________________________________________________________________________



### 2️⃣ Análisis exploratorio inicial

In [42]:
# Función para realizar un análisis exploratorio inicial de los DataFrames.

def eda(df, nombre):
    print(f"***** {nombre} *****")
    print("_" * 50)
    print()
    print("Info:")
    display(df.info())
    print("_" * 50)
    print()
    print("Shape:", df.shape)
    print("_" * 50)
    print()
    print("Columnas:", list(df.columns))
    print("_" * 50)
    print()
    print("Tipos de datos (dtypes):")
    print(df.dtypes)
    print("_" * 50)
    print()
    print("\nNulos por columna:")
    print(df.isna().sum().to_frame("nulos"))
    print("_" * 50)
    print()
    print("\nPrimeras filas:")
    display(df.head(4))
    print("_" * 50)
    print()
    print("\nUltimas filas:")
    display(df.tail(4))
    print("_" * 50)
    print()
    print("\nDescribe (numérico):")
    display(df.describe(include='number').round(2))
    print("_"*80)

In [43]:
# *************  EDA VENTAS  **************

eda(df_ventas,"Ventas (EDA)")

***** Ventas (EDA) *****
__________________________________________________

Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   object 
 5   categoria    3035 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 142.4+ KB


None

__________________________________________________

Shape: (3035, 6)
__________________________________________________

Columnas: ['id_venta', 'producto', 'precio', 'cantidad', 'fecha_venta', 'categoria']
__________________________________________________

Tipos de datos (dtypes):
id_venta         int64
producto        object
precio          object
cantidad       float64
fecha_venta     object
categoria       object
dtype: object
__________________________________________________


Nulos por columna:
             nulos
id_venta         0
producto         0
precio           2
cantidad         2
fecha_venta      0
categoria        0
__________________________________________________


Primeras filas:


id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro decorativo   $69.94       5.0  02/01/2024   
1       811    Lámpara de mesa  $105.10       5.0  02/01/2024   
2      1156           Secadora   $97.96       3.0  02/01/2024   
3      1372           Heladera  $114.35       8.0  02/01/2024   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos

__________________________________________________


Ultimas filas:


id_venta                producto   precio  cantidad fecha_venta  \
3031      2276                  Laptop   $85.27       9.0  30/12/2024   
3032      2696                  Laptop  $107.81       4.0  30/12/2024   
3033      2913              Smartphone   $99.85       7.0  30/12/2024   
3034      2930  Consola de videojuegos   $55.47       6.0  30/12/2024   

        categoria  
3031  Electrónica  
3032  Electrónica  
3033  Electrónica  
3034  Electrónica

__________________________________________________


Describe (numérico):


id_venta  cantidad
count   3035.00   3033.00
mean    1499.85      6.50
std      866.47      3.46
min        1.00      1.00
25%      748.50      3.00
50%     1502.00      7.00
75%     2249.50      9.00
max     3000.00     12.00

________________________________________________________________________________


### ***Resumen***

Al realizar el análisis exploratorio de **Ventas.csv** se observa:


1.   El dataframe posee 3035 filas y 6 columnas
2.   La columna Precio es object y debería ser flotante (lleva el signo $ delante del número)
1.   La columna Fecha es object y debería ser datetime
4.   Hay cuatro campos nulos (dos en precio y dos en cantidad)





In [44]:
# EDA MARKETING

eda(df_marketing,"Marketing (EDA)")

***** Marketing (EDA) *****
__________________________________________________

Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id_campanha   90 non-null     int64  
 1   producto      90 non-null     object 
 2   canal         90 non-null     object 
 3   costo         90 non-null     float64
 4   fecha_inicio  90 non-null     object 
 5   fecha_fin     90 non-null     object 
dtypes: float64(1), int64(1), object(4)
memory usage: 4.3+ KB


None

__________________________________________________

Shape: (90, 6)
__________________________________________________

Columnas: ['id_campanha', 'producto', 'canal', 'costo', 'fecha_inicio', 'fecha_fin']
__________________________________________________

Tipos de datos (dtypes):
id_campanha       int64
producto         object
canal            object
costo           float64
fecha_inicio     object
fecha_fin        object
dtype: object
__________________________________________________


Nulos por columna:
              nulos
id_campanha       0
producto          0
canal             0
costo             0
fecha_inicio      0
fecha_fin         0
__________________________________________________


Primeras filas:


id_campanha         producto  canal  costo fecha_inicio   fecha_fin
0           74  Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12           Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32  Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21       Smartphone   RRSS   6.37   29/03/2024  16/05/2024

__________________________________________________


Ultimas filas:


id_campanha            producto  canal  costo fecha_inicio  fecha_fin
86           89           Televisor     TV   4.98   13/12/2024   8/2/2025
87           68   Rincón de plantas     TV   5.81   17/12/2024  14/2/2025
88           33            Secadora  Email   3.80   20/12/2024   7/1/2025
89           11  Freidora eléctrica   RRSS   5.27   29/12/2024  21/1/2025

__________________________________________________


Describe (numérico):


id_campanha  costo
count        90.00  90.00
mean         45.50   4.93
std          26.12   0.95
min           1.00   2.95
25%          23.25   4.37
50%          45.50   4.90
75%          67.75   5.56
max          90.00   7.39

________________________________________________________________________________


### ***Resumen***

Al realizar el análisis exploratorio de **Marketing.csv** se observa:


1.   El dataframe posee 90 filas y 6 columnas
1.   Las columnas fecha_inicio y fecha_fin son object y deberían ser datetime
4.   No hay campos nulos

In [45]:
# EDA CLIENTES
eda(df_clientes,"Clientes (EDA)")

***** Clientes (EDA) *****
__________________________________________________

Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 567 entries, 0 to 566
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id_cliente  567 non-null    int64  
 1   nombre      567 non-null    object 
 2   edad        567 non-null    int64  
 3   ciudad      567 non-null    object 
 4   ingresos    567 non-null    float64
dtypes: float64(1), int64(2), object(2)
memory usage: 22.3+ KB


None

__________________________________________________

Shape: (567, 5)
__________________________________________________

Columnas: ['id_cliente', 'nombre', 'edad', 'ciudad', 'ingresos']
__________________________________________________

Tipos de datos (dtypes):
id_cliente      int64
nombre         object
edad            int64
ciudad         object
ingresos      float64
dtype: object
__________________________________________________


Nulos por columna:
            nulos
id_cliente      0
nombre          0
edad            0
ciudad          0
ingresos        0
__________________________________________________


Primeras filas:


id_cliente               nombre  edad         ciudad  ingresos
0           1      Aloysia Screase    44  Mar del Plata  42294.68
1           2  Kristina Scaplehorn    25        Posadas  24735.04
2           3       Filip Castagne    50    Resistencia  35744.85
3           4          Liuka Luard    39   Bahía Blanca  27647.96

__________________________________________________


Ultimas filas:


id_cliente           nombre  edad       ciudad  ingresos
563         564      Fleming Gow    39     Santa Fe  43674.96
564         565  Jewelle Mabbett    33      Córdoba  30522.64
565         566      Lauri Munns    23  Resistencia  31259.14
566         567      Micah Matis    31   Corrientes  42927.86

__________________________________________________


Describe (numérico):


id_cliente    edad  ingresos
count      567.00  567.00    567.00
mean       284.00   37.94  34668.74
std        163.82   10.20  12974.53
min          1.00   20.00    170.29
25%        142.50   30.00  26015.24
50%        284.00   37.00  35066.83
75%        425.50   43.00  42457.10
max        567.00   81.00  88053.01

________________________________________________________________________________


### ***Resumen***

Al realizar el análisis exploratorio de **Clientes.csv** se observa:


1.   El dataframe posee 567 filas y 6 columnas
1.   Las columnas poseen el tipo de dato correcto.
4.   No hay campos nulos

### 3️⃣ Calidad de los datos

In [46]:
# Función para identificar los duplicados

def duplicados(df, nombre,clave=None):
   print(f"***** Duplicados en {nombre} *****")
   print("_" * 50)
   print()
   dup_rows = df.duplicated(keep=False).sum()
   print("Filas duplicadas (exactas):", dup_rows)
   if clave and clave in df.columns:
        dup_key = df[clave].duplicated(keep=False).sum()
        print(f"Duplicados por clave '{clave}':", dup_key)
        if dup_key > 0:
             duplicados_ordenados = (
                df[df[clave].duplicated(keep=False)][clave]
                .value_counts()                # Cuenta cuántas veces aparece cada valor
                .sort_values(ascending=False)   # Ordena de mayor a menor (más duplicados arriba)
              )
             print("\n🔁 Top valores duplicados más frecuentes:")
             # Mostrar solo los primeros 10 (los más repetidos)
             display(duplicados_ordenados.head(10))
        else:
            print(f"No se encontraron duplicados en la clave '{clave}'.")
   else:
        # Si la clave no fue pasada o no existe en el DataFrame
        if clave:
            print(f"La clave '{clave}' no existe en el DataFrame.")
        else:
            print("No se indicó una clave para analizar duplicados por columna.")

In [47]:
duplicados(df_ventas,"Ventas",clave="id_venta")

***** Duplicados en Ventas *****
__________________________________________________

Filas duplicadas (exactas): 70
Duplicados por clave 'id_venta': 70

🔁 Top valores duplicados más frecuentes:


id_venta
56      2
421     2
424     2
1868    2
2545    2
2778    2
145     2
300     2
439     2
906     2
Name: count, dtype: int64

In [48]:
duplicados(df_marketing,"Marketing",clave="id_campanha")


***** Duplicados en Marketing *****
__________________________________________________

Filas duplicadas (exactas): 0
Duplicados por clave 'id_campanha': 0
No se encontraron duplicados en la clave 'id_campanha'.


In [49]:
duplicados(df_clientes,"Clientes",clave="id_cliente")

***** Duplicados en Clientes *****
__________________________________________________

Filas duplicadas (exactas): 0
Duplicados por clave 'id_cliente': 0
No se encontraron duplicados en la clave 'id_cliente'.


## 🧹 Etapa 2: Preprocesamiento y Limpieza de Datos
**Objetivo:** Demostrar conocimiento de las técnicas de limpieza y transformación de datos.

### 4️⃣ Limpieza de datos

In [50]:
# Limpio el conjunto de datos eliminando duplicados y caracteres no deseados.

# Primero creo copias independientes para no modificar los df originales

df_ventas_clean = df_ventas.copy()
df_marketing_clean = df_marketing.copy()
df_clientes_clean = df_clientes.copy()

In [51]:
# Elimino filas duplicadas completas en cada df

df_ventas_clean = df_ventas_clean.drop_duplicates()
df_marketing_clean = df_marketing_clean.drop_duplicates()
df_clientes_clean = df_clientes_clean.drop_duplicates()



In [52]:
# Verifico que ya no existan duplicados en los df_clean

print()
duplicados(df_ventas_clean,"Ventas",clave="id_venta")
print("Shape:  ", df_ventas_clean.shape)
print("_" * 50)
print()
duplicados(df_marketing_clean,"Marketing",clave="id_campanha")
print("Shape:  ", df_marketing_clean.shape)
print("_" * 50)
print()
duplicados(df_clientes_clean,"Clientes",clave="id_cliente")
print("Shape:  ", df_clientes_clean.shape)
print("_" * 50)


***** Duplicados en Ventas *****
__________________________________________________

Filas duplicadas (exactas): 0
Duplicados por clave 'id_venta': 0
No se encontraron duplicados en la clave 'id_venta'.
Shape:   (3000, 6)
__________________________________________________

***** Duplicados en Marketing *****
__________________________________________________

Filas duplicadas (exactas): 0
Duplicados por clave 'id_campanha': 0
No se encontraron duplicados en la clave 'id_campanha'.
Shape:   (90, 6)
__________________________________________________

***** Duplicados en Clientes *****
__________________________________________________

Filas duplicadas (exactas): 0
Duplicados por clave 'id_cliente': 0
No se encontraron duplicados en la clave 'id_cliente'.
Shape:   (567, 5)
__________________________________________________


In [53]:
# Normalizo fechas

def fecha_norm (df):
  flag=True
  for col in df.columns:
        if "fecha" in col.lower():  # detecta columnas con la palabra "fecha"
            df[col] = pd.to_datetime(df[col], errors="coerce", dayfirst=True)
            print(f"Columna '{col}' normalizada a '{df[col].dtypes}'")
            flag=False
            print("-" * 50)

  if flag:
      print("No se encontraron columnas con el patrón 'fecha' en el DataFrame.")


In [54]:
fecha_norm(df_ventas_clean)


Columna 'fecha_venta' normalizada a 'datetime64[ns]'
--------------------------------------------------


In [55]:
fecha_norm(df_marketing_clean)

Columna 'fecha_inicio' normalizada a 'datetime64[ns]'
--------------------------------------------------
Columna 'fecha_fin' normalizada a 'datetime64[ns]'
--------------------------------------------------


In [56]:
fecha_norm(df_clientes_clean)

No se encontraron columnas con el patrón 'fecha' en el DataFrame.


In [57]:
# Función para normalizar texto

def normalizar_texto(df):
    for col in df.select_dtypes(include="object").columns:
        df[col] = (                                 # Usar () cuando va en distintas lineas
            df[col]
            .astype(str)                              # Convierte cualquier tipo a string
            .str.strip()                               # Elimina espacios al inicio y final
            .str.replace(r"[\u200b\t\r\n]", "", regex=True) # Borra tabs, saltos, etc
            .str.replace(" +", " ", regex=True)             # Reemplaza espacios consecutivos x 1
            .str.title()                               # Convierte a Título
        )

        #df[col] = df[col].astype(str).str.strip().str.replace(r"[\u200b\t\r\n]", "", regex=True).str.replace(" +", " ", regex=True).str.title()
    return df


In [66]:
df_ventas_clean = normalizar_texto(df_ventas_clean)
print("-" * 100)
print("Texto normalizado en df_ventas_clean")
df_marketing_clean = normalizar_texto(df_marketing_clean)
print("-" * 100)
print("Texto normalizado en df_marketing_clean")
df_clientes_clean = normalizar_texto(df_clientes_clean)
print("-" * 100)
print("Texto normalizado en df_clientes_clean")

----------------------------------------------------------------------------------------------------
Texto normalizado en df_ventas_clean
----------------------------------------------------------------------------------------------------
Texto normalizado en df_marketing_clean
----------------------------------------------------------------------------------------------------
Texto normalizado en df_clientes_clean


In [67]:
# Normalizo campos numéricos

# Columnas de PRECIO

def normalizar_precio(df):
  for col in df.columns:
    flag=True
    if "precio" in col.lower():
      df[col]=df[col].astype(str)
      df[col]=df[col].str.replace("$","",regex=False)
      df[col]=df[col].str.replace(",","",regex=False)
      df[col]=df[col].str.strip()
      df[col]=pd.to_numeric(df[col],errors="coerce")
      print("Precio normalizado en df_ventas_clean")
      flag=False
      print("-" * 50)

  if flag:
      print("No se encontraron columnas con el patrón 'precio' en el DataFrame.")
      print("-" * 50)

  return df



  # CANTIDAD / Cambia de flotante a entero

def normalizar_flot_a_int(df,col):
  #df[col]=df[col].astype(str)
  #df[col]=df[col].str.replace(".0","",regex=False)
  df[col]=pd.to_numeric(df[col],errors="coerce").astype("Int64")
  return df
#
  #if "cantidad" in ventas_clean.columns:
    #ventas_clean["cantidad"] = pd.to_numeric(
     #   ventas_clean["cantidad"], errors="coerce"
    #).astype("Int64")
    # .astype("Int64") usa el tipo entero de pandas que permite valores nulos (NaN)

In [68]:
df_ventas_clean = normalizar_precio(df_ventas_clean)
df_ventas_clean = normalizar_flot_a_int(df_ventas_clean,"cantidad")
print("-" * 100)


Precio normalizado en df_ventas_clean
--------------------------------------------------
No se encontraron columnas con el patrón 'precio' en el DataFrame.
--------------------------------------------------
----------------------------------------------------------------------------------------------------


In [61]:
# Verifico que los tipos de datos esten correctos en cada dataframe.

print("\n****** Archivo ventas.csv ******")
print("Shape :",df_ventas_clean.shape)
print("Dtypes:")
print(df_ventas_clean.dtypes)
display(df_ventas_clean.head())
print()
print("_"*80)
print("\n****** Archivo marketing.csv ******")
print("Shape :",df_marketing_clean.shape)
print("Dtypes:")
print(df_marketing_clean.dtypes)
display(df_marketing_clean.head())
print()
print("_"*80)
print("\n****** Archivo clientes.csv ******")
print("Shape :",df_clientes_clean.shape)
print("Dtypes:")
print(df_clientes_clean.dtypes)
display(df_clientes_clean.head())
print("_"*80)


****** Archivo ventas.csv ******
Shape : (3000, 6)
Dtypes:
id_venta                int64
producto               object
precio                float64
cantidad                Int64
fecha_venta    datetime64[ns]
categoria              object
dtype: object


id_venta           producto  precio  cantidad fecha_venta  \
0       792  Cuadro Decorativo   69.94         5  2024-01-02   
1       811    Lámpara De Mesa  105.10         5  2024-01-02   
2      1156           Secadora   97.96         3  2024-01-02   
3      1372           Heladera  114.35         8  2024-01-02   
4      1546           Secadora  106.21         4  2024-01-02   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos


________________________________________________________________________________

****** Archivo marketing.csv ******
Shape : (90, 6)
Dtypes:
id_campanha              int64
producto                object
canal                   object
costo                  float64
fecha_inicio    datetime64[ns]
fecha_fin       datetime64[ns]
dtype: object


id_campanha         producto  canal  costo fecha_inicio  fecha_fin
0           74  Adorno De Pared     Tv   4.81   2024-03-20 2024-05-03
1           12           Tablet   Rrss   3.40   2024-03-26 2024-05-13
2           32  Lámpara De Mesa  Email   5.54   2024-03-28 2024-04-20
3           21       Smartphone   Rrss   6.37   2024-03-29 2024-05-16
4           58         Alfombra  Email   4.25   2024-03-31 2024-05-05


________________________________________________________________________________

****** Archivo clientes.csv ******
Shape : (567, 5)
Dtypes:
id_cliente      int64
nombre         object
edad            int64
ciudad         object
ingresos      float64
dtype: object


id_cliente               nombre  edad         ciudad  ingresos
0           1      Aloysia Screase    44  Mar Del Plata  42294.68
1           2  Kristina Scaplehorn    25        Posadas  24735.04
2           3       Filip Castagne    50    Resistencia  35744.85
3           4          Liuka Luard    39   Bahía Blanca  27647.96
4           5        Dore Cockshtt    28        Rosario  28245.65

________________________________________________________________________________


In [62]:
# Guardo los DF en archivos .csv

df_ventas_clean.info()
df_ventas_clean.to_csv("/content/drive/MyDrive/datasets/ventas_clean.csv", index=False)
df_clientes_clean.to_csv("/content/drive/MyDrive/datasets/clientes_clean.csv", index=False)
df_marketing_clean.to_csv("/content/drive/MyDrive/datasets/marketing_clean.csv", index=False)

print("✅ Archivos guardados: ventas_clean.csv, clientes_clean.csv, marketing_clean.csv")

<class 'pandas.core.frame.DataFrame'>
Index: 3000 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   id_venta     3000 non-null   int64         
 1   producto     3000 non-null   object        
 2   precio       2998 non-null   float64       
 3   cantidad     2998 non-null   Int64         
 4   fecha_venta  3000 non-null   datetime64[ns]
 5   categoria    3000 non-null   object        
dtypes: Int64(1), datetime64[ns](1), float64(1), int64(1), object(2)
memory usage: 167.0+ KB
✅ Archivos guardados: ventas_clean.csv, clientes_clean.csv, marketing_clean.csv


### 5️⃣ Transformación de datos

In [63]:
# Crear una tabla de ventas que muestre solo los productos con alto rendimiento.
# Calcular el percentil 80
# Filtrar los productos que superen ese umbral en ventas.

# Calculo el ingreso y creo un nuevo df con la columna "ingreso"
df_ventas_ingresos = df_ventas_clean.assign(ingreso = df_ventas_clean["precio"] * df_ventas_clean["cantidad"])

# Agrupo filas y obtengo métricas
resumen_prod = (
    df_ventas_ingresos.groupby(by="producto",dropna=False,as_index=False,observed=False)
    .agg(
        ingreso_total=('ingreso', 'sum'),
        unidades=('cantidad', 'sum'),
        precio_promedio=('precio', 'mean'),
        registros=('ingreso', 'size'))
)

# Redondeo el 'precio_promedio'
resumen_prod['precio_promedio'] = resumen_prod['precio_promedio'].round(2)

#obtengo percentil 80
p80_ingreso=resumen_prod["ingreso_total"].quantile(q=0.8,interpolation="linear")
#print("Percentil 80: ",p80_ingreso.round(2))

# Filtro los productos de alto rendimiento y ordeno de mayor a menor
prod_alto_rendimiento = (
    resumen_prod.query("ingreso_total >= @p80_ingreso",engine="python")
    .sort_values(by=["ingreso_total","unidades"],ascending=[False,False],ignore_index=True)
)

# Muestro resultados
print()
print("**  Columna: Producto  **")
print(f"Umbral (percentil 80) de ingreso_total: {float(p80_ingreso):,.2f}")
print("✅ Productos de ALTO RENDIMIENTO (top 20% por ingreso):")
display(prod_alto_rendimiento.head(20))


**  Columna: Producto  **
Umbral (percentil 80) de ingreso_total: 52,518.85
✅ Productos de ALTO RENDIMIENTO (top 20% por ingreso):


producto  ingreso_total  unidades  precio_promedio  registros
0    Lámpara De Mesa       82276.38      1112            72.72        176
1        Auriculares       74175.58       958            76.30        143
2         Microondas       72562.89       912            79.18        135
3           Cafetera       59607.31       765            79.05        117
4  Cuadro Decorativo        54297.6       726            74.58        100
5         Smartphone       54132.44       665            81.40        101

### 6️⃣ Agregación

In [64]:
# TODO: Resumir las ventas por categoría de producto y analizar los ingresos generados.
# Sugerencia: usar .groupby() y .agg() para generar métricas como suma y promedio.

# Calculo el ingreso y creo un nuevo df con la columna "ingreso"
df_ventas_cat = df_ventas_clean.assign(ingreso = df_ventas_clean["precio"] * df_ventas_clean["cantidad"])

# Agrupo filas y obtengo métricas
resumen_cat = (
    df_ventas_cat.groupby(by="categoria",dropna=False,as_index=False)
    .agg(
        ingreso_total=('ingreso', 'sum'),
        unidades=('cantidad', 'sum'),
        precio_promedio=('precio', 'mean'),
        registros=('ingreso', 'size')
        )
    .sort_values(
        by='ingreso_total',
        ascending=False,
        na_position='last',
        ignore_index=True
    )
)

# Redondeo el 'precio_promedio'
resumen_cat['precio_promedio'] = resumen_cat['precio_promedio'].round(2)


# Muestro resultados
print()
print("**  Columna: Categoría  **")
print("Resumen por categoría (ordenado por ingreso_total):")
display(resumen_cat.head(20))


**  Columna: Categoría  **
Resumen por categoría (ordenado por ingreso_total):


categoria  ingreso_total  unidades  precio_promedio  registros
0  Electrodomésticos      505299.63      6592            76.52       1000
1        Electrónica       482577.8      6413            75.25        999
2         Decoración      479216.09      6490            74.10       1001

### 7️⃣ Integración de datos, opcional, NO OBLIGATORIO

In [65]:
# TODO: Combinar los sets de datos de ventas y marketing para obtener una visión más amplia de las tendencias.
# Sugerencia: usar pd.merge() especificando la clave común entre ambos DataFrames.
# Documentar cualquier observación relevante sobre la combinación de datos.
